src/drivers/intel/ssp/ssp.c

```diff
@@ -766,19 +780,80 @@ static int ssp_get_hw_params(struct dai *dai,
 	return 0;
 }
 
-/* SSP hw params */
+/*
+ * enabled clock MCLK/BCLK to support codec that need MCLK/BCLK
+ * This is called before trigger(COMP_TRIGGER_START)
+ */
 static int ssp_hw_params(struct dai *dai,
 			 struct sof_ipc_stream_params  *params)
 {
-	dai_info(dai, "ssp_hw_params()");
+	struct ssp_pdata *ssp = dai_get_drvdata(dai);
+	int ret;
 
+	dai_info(dai, "ssp_hw_params() params.clks_control 0x%x", ssp->params.clks_control);
+
+	spin_lock(&dai->lock);
+
+	if ((ssp->params.clks_control &
+		 SOF_DAI_INTEL_SSP_CLKCTRL_MCLK_ES)) {
+		ret = ssp_set_mclk(dai);
+		if (ret < 0)
+			goto out;
+	}
+
+	if ((ssp->params.clks_control &
+		 SOF_DAI_INTEL_SSP_CLKCTRL_BCLK_ES)) {
+		ret = ssp_set_bclk(dai);
+		if (ret < 0)
+			goto out;
+		/* enable port */
+		ssp_update_bits(dai, SSCR0, SSCR0_SSE, SSCR0_SSE);
+	}
+out:
+
+	spin_unlock(&dai->lock);
 	return 0;
 }
 
-/* SSP hw free */
+/*
+ * disabled clock MCLK/BCLK to support codec that need MCLK/BCLK
+ * This is called after trigger(COMP_TRIGGER_STOP)
+ */
 static void ssp_hw_free(struct dai *dai)
 {
+	struct ssp_pdata *ssp = dai_get_drvdata(dai);
+
 	dai_info(dai, "ssp_hw_free()");
+
+	spin_lock(&dai->lock);
+
+	if (ssp->state[SOF_IPC_STREAM_CAPTURE] == COMP_STATE_PREPARE &&
+	    ssp->state[SOF_IPC_STREAM_PLAYBACK] == COMP_STATE_PREPARE) {
+		if ((ssp->params.clks_control &
+			 SOF_DAI_INTEL_SSP_CLKCTRL_BCLK_ES)) {
+			ssp_update_bits(dai, SSCR0, SSCR0_SSE, 0);
+			dai_info(dai, "ssp_hw_free(), SSP port disabled");
+		}
+	}
+
+	/* release clocks if SSP is inactive */
+	if (ssp->state[SOF_IPC_STREAM_PLAYBACK] != COMP_STATE_ACTIVE &&
+	    ssp->state[SOF_IPC_STREAM_CAPTURE] != COMP_STATE_ACTIVE) {
```

Comment:
why use == PREPARE and != ACTIVE is the two different bclk and mclk cases? the state management isn't clear to me.

Reply:
> why use == PREPARE and != ACTIVE is the two different bclk and mclk cases? the state management isn't clear to me.

That was copied from `ssp_post_stop()`. I will change it to `== COMP_STATE_PREPARE`. I also thing using `== COMP_STATE_ACTIVE` to test if clock is already active is not right. The clock will keep active when ssp->state = COMP_STATE_PAUSED. @keyonjie what do you think?